### PR TITLE
Avoid unintentional unsigned integer wraparounds

### PR DESCRIPTION
**Pull request by practicalswift**
Avoid unintentional unsigned integer wraparounds.

### CODE REVIEW

File: src/test/prevector_tests.cpp [hunk 1]
```diff
@@ -206,7 +206,7 @@ BOOST_AUTO_TEST_CASE(PrevectorTestInt)
                 test.erase(InsecureRandRange(test.size()));
             }
             if (InsecureRandBits(3) == 2) {
-                int new_size = std::max<int>(0, std::min<int>(30, test.size() + (InsecureRandRange(5)) - 2));
+                int new_size = std::max<int>(0, std::min<int>(30, static_cast<int>(test.size()) + static_cast<int>(InsecureRandRange(5)) - 2));
```
Comment (laanwj) [inline src/test/prevector_tests.cpp]: Does this still need the ```<int>``` qualifier on `std::max/min` if you cast all the arguments to int?

**Comment by practicalswift** [comment]
@laanwj Good point. Fixed! :-)

**Comment by practicalswift** [comment]
Added commits from #11547 ("Avoid unintended unsigned integer wraparounds in FormatScript(...) and SplitHostPort(...)") as requested :-)

**Comment by practicalswift** [comment]
Added a few more wrap-arounds and squashed. Now at 15 fixed wrap-arounds.

**Comment by sipa** [comment]
We tend to use C-style casts for primitive types... just for brevity.

**Comment by practicalswift** [comment]
@sipa I'll change! Other than that, do the changes look reasonable? :-)

File: src/policy/fees.cpp [hunk 2]
```diff
@@ -291,7 +291,7 @@ double TxConfirmStats::EstimateMedianVal(int confTarget, double sufficientTxVal,
         totalNum += txCtAvg[bucket];
         failNum += failAvg[periodTarget - 1][bucket];
         for (unsigned int confct = confTarget; confct < GetMaxConfirms(); confct++)
-            extraNum += unconfTxs[(nBlockHeight - confct)%bins][bucket];
+            extraNum += unconfTxs[(nBlockHeight >= confct ? nBlockHeight - confct : 0) % bins][bucket];
```
Comment (practicalswift) [inline src/policy/fees.cpp]: This change needs extra review. Is this the correct way to handle this wrap-around?

**Comment by practicalswift** [comment]
@sipa Now using C-style casts for primitive types :-)

**Comment by practicalswift** [comment]
Added another wraparound fix (this time in `AcceptBlock(…)`) and squashed.

Anyone willing to review? :-)

**Comment by practicalswift** [comment]
Rebased! :-)

**Comment by practicalswift** [comment]
Anyone willing to review - ACK or NACK? :-)

**Comment by practicalswift** [comment]
Ping? :-)

**Comment by practicalswift** [comment]
Do we not care about integer wrap-arounds? If so let me know and I'll close :-)

**Comment by MarcoFalke** [comment]
ACK test changes, they seem straightforward. The other need a cautious review, since they change behavior.

File: src/test/scheduler_tests.cpp [hunk 3]
```diff
@@ -56,8 +56,8 @@ BOOST_AUTO_TEST_CASE(manythreads)
     int counter[10] = { 0 };
     FastRandomContext rng(42);
     auto zeroToNine = [](FastRandomContext& rc) -> int { return rc.randrange(10); }; // [0, 9]
-    auto randomMsec = [](FastRandomContext& rc) -> int { return -11 + rc.randrange(1012); }; // [-11, 1000]
-    auto randomDelta = [](FastRandomContext& rc) -> int { return -1000 + rc.randrange(2001); }; // [-1000, 1000]
+    auto randomMsec = [](FastRandomContext& rc) -> int { return -11 + (int)rc.randrange(1012); }; // [-11, 1000]
+    auto randomDelta = [](FastRandomContext& rc) -> int { return -1000 + (int)rc.randrange(2001); }; // [-1000, 1000]
```
Comment (MarcoFalke) [inline src/test/scheduler_tests.cpp]: This change is already in master

File: src/txmempool.cpp [hunk 4]
```diff
@@ -221,7 +221,7 @@ void CTxMemPool::UpdateAncestorsOf(bool add, txiter it, setEntries &setAncestors
         UpdateChild(piter, it, add);
     }
     const int64_t updateCount = (add ? 1 : -1);
-    const int64_t updateSize = updateCount * it->GetTxSize();
+    const int64_t updateSize = updateCount * (int64_t)it->GetTxSize();
```
Comment (promag) [inline src/txmempool.cpp]: Why is this necessary?

File: src/core_write.cpp [hunk 5]
```diff
@@ -58,7 +58,7 @@ std::string FormatScript(const CScript& script)
         ret += strprintf("0x%x ", HexStr(it2, script.end()));
         break;
     }
-    return ret.substr(0, ret.size() - 1);
+    return ret.size() > 0 ? ret.substr(0, ret.size() - 1) : ret;
```
Comment (promag) [inline src/core_write.cpp]: IMO there is no need in copying the whole string. Suggestion:
```cpp
if (!ret.empty()) ret.pop_back();
return ret;
```

File: src/policy/fees.cpp (hunk 2 above)
Comment (promag) [reply src/policy/fees.cpp]: Does it matters if `nBlockHeight < GetMaxConfirms()`? Maybe we should just skip `EstimateMedianVal` entirely if so. cc @morcos

Comment (morcos) [reply src/policy/fees.cpp]: Actually I think the existing code is broken, we should fix it by replacing the for loop with:
`for (unsigned int confct = confTarget; confct < GetMaxConfirms() && confct <= nBlockHeight; confct++)`

Comment (promag) [reply src/policy/fees.cpp]: That makes sense.

Comment (practicalswift) [reply src/policy/fees.cpp]: Thanks @promag and @morcos. Now updated. Please re-review :-)

**Comment by morcos** [comment]
Re: policy/fees.cpp changes   utACK   (didn't review anything else)

Thanks!

File: src/validation.cpp [hunk 6]
```diff
@@ -1588,7 +1588,8 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
                 error("DisconnectBlock(): transaction and undo data inconsistent");
                 return DISCONNECT_FAILED;
             }
-            for (unsigned int j = tx.vin.size(); j-- > 0;) {
+            for (unsigned int j = tx.vin.size(); j > 0;) {
```
Comment (promag) [inline src/validation.cpp]: It doesn't matter if it wraps, `j` is only used inside the loop and it only wraps when `j = 0` which is when the loop ends.

Anyway, how about changing to:
```cpp
for (int j = tx.vin.size() - 1; j >= 0; --j) {
```
which IMO is more clear.

File: src/txmempool.cpp (hunk 4 above)
Comment (practicalswift) [reply src/txmempool.cpp]: Consider the case where `add == false` and `it->GetTxSize() == 2`.

Then an unsigned integer overflow occurs when calculating `updateCount * it->GetTxSize()` (but not for `updateCount * (int64_t)it->GetTxSize()`).

**Comment by practicalswift** [comment]
@promag Thanks for reviewing. Switched to suggested `for` formulation. Please re-review.

File: src/validation.cpp (hunk 6 above)
Comment (practicalswift) [reply src/validation.cpp]: All unsigned integer overflows matter :-)

File: src/chain.h [hunk 7]
```diff
@@ -475,7 +475,7 @@ class CChain {
 
     /** Return the maximal height in the chain. Is equal to chain.Tip() ? chain.Tip()->nHeight : -1. */
     int Height() const {
-        return vChain.size() - 1;
+        return (int)vChain.size() - 1;
```
Comment (promag) [inline src/chain.h]: Why is this cast necessary?

Comment (practicalswift) [reply src/chain.h]: An unsigned integer overflow occurs in the case where `vChain.size() == 0`.

(And yes, the value return by `Height()` will be the expected `-1`, but an overflow occurred before returning :-))

**Comment by practicalswift** [comment]
Anyone willing to review?

Getting the project to compile cleanly under `-fsanitize=integer` would be very nice :-)

File: src/policy/fees.cpp [hunk 8]
```diff
@@ -288,7 +288,7 @@ double TxConfirmStats::EstimateMedianVal(int confTarget, double sufficientTxVal,
         nConf += confAvg[periodTarget - 1][bucket];
         totalNum += txCtAvg[bucket];
         failNum += failAvg[periodTarget - 1][bucket];
-        for (unsigned int confct = confTarget; confct < GetMaxConfirms(); confct++)
+        for (unsigned int confct = confTarget; confct < GetMaxConfirms() && confct <= nBlockHeight; confct++)
```
Comment (promag) [inline src/policy/fees.cpp]: Nit, `++ confct`.

File: src/txmempool.cpp (hunk 4 above)
Comment (promag) [reply src/txmempool.cpp]: :+1:

File: src/chain.h (hunk 7 above)
Comment (promag) [reply src/chain.h]: :+1:

File: src/validation.cpp [hunk 9]
```diff
@@ -1588,7 +1588,7 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
                 error("DisconnectBlock(): transaction and undo data inconsistent");
                 return DISCONNECT_FAILED;
             }
-            for (unsigned int j = tx.vin.size(); j-- > 0;) {
+            for (int j = tx.vin.size() - 1; j >= 0; --j) {
```
Comment (promag) [inline src/validation.cpp]: Should cast here too (`int j = (int)tx.vin.size() - 1`)?

**Comment by practicalswift** [comment]
@promag Thanks for reviewing! Feedback addressed!

@sipa @promag Please re-review :-)

File: src/policy/fees.cpp [hunk 10]
```diff
@@ -288,7 +288,7 @@ double TxConfirmStats::EstimateMedianVal(int confTarget, double sufficientTxVal,
         nConf += confAvg[periodTarget - 1][bucket];
         totalNum += txCtAvg[bucket];
         failNum += failAvg[periodTarget - 1][bucket];
-        for (unsigned int confct = confTarget; confct < GetMaxConfirms(); confct++)
+        for (unsigned int confct = confTarget; confct < GetMaxConfirms() && confct <= nBlockHeight; ++confct)
```
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/policy/fees.cpp]: Good point! Fixed :-)

**Comment by practicalswift** [comment]
@arvidn Updated to address your feedback. Please re-review!

BTW, I love your talk about C++ integers! Great to have you as a reviewer of this PR :-)

**Comment by practicalswift** [comment]
Updated! Please re-review.

File: src/policy/fees.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/policy/fees.cpp]: @arvidn Thanks for reviewing! Do you suggest changing it from `unsigned int` to `int` or leaving it as is? :-)  Note that `GetMaxConfirms()` returns an `unsigned int`.

[the remark this answers is not in the document]
Comment (arvidn) [reply src/policy/fees.cpp]: yes, in general I would recommend using signed integers for anything that is expected to have normal arithmetic (as opposed to modulo-2 arithmetic). i.e. `GetMaxConfirms()` should probably return a signed integer. But I don't think it's practical to "unravel" all such dependencies in one commit.

[the remark this answers is not in the document]
Comment (arvidn) [reply src/policy/fees.cpp]: yes, in general I would recommend using signed integers for anything that is expected to have normal arithmetic (as opposed to modulo-2 arithmetic). i.e. `GetMaxConfirms()` should probably return a signed integer. But I don't think it's practical to "unravel" all such dependencies in one commit.

File: src/script/interpreter.cpp [hunk 11]
```diff
@@ -51,8 +51,8 @@ bool CastToBool(const valtype& vch)
  * Script is a stack machine (like Forth) that evaluates a predicate
  * returning a bool indicating valid or not.  There are no loops.
  */
-#define stacktop(i)  (stack.at(stack.size()+(i)))
-#define altstacktop(i)  (altstack.at(altstack.size()+(i)))
+#define stacktop(i)  (stack.at((size_t)((ssize_t)stack.size() + (ssize_t)i)))
```
Comment (practicalswift) [inline src/script/interpreter.cpp]:
```
2018-09-19 12:07:43 clang-tidy(pr=11535): src/script/interpreter.cpp:54:74: warning: macro argument should be enclosed in parentheses [misc-macro-parentheses]
```

File: src/script/interpreter.cpp [hunk 12]
```diff
@@ -51,8 +51,8 @@ bool CastToBool(const valtype& vch)
...
+#define altstacktop(i)  (altstack.at((size_t)((ssize_t)altstack.size() + (ssize_t)i)))
```
Comment (practicalswift) [inline src/script/interpreter.cpp]:
```
2018-09-19 12:07:43 clang-tidy(pr=11535): src/script/interpreter.cpp:55:83: warning: macro argument should be enclosed in parentheses [misc-macro-parentheses]
```

**Comment by practicalswift** [comment]
To increase the likelihood of this PR getting merged – should I limit it to a subset that is low-risk/non-controversial? It would be really nice to get rid of these unintended unsigned integer wraparounds. Please advice :-)

**Comment by flack** [comment]
> Please help me find that subset :-)

@practicalswift see https://github.com/bitcoin/bitcoin/pull/11535#issuecomment-367844106

**Comment by arvidn** [comment]
I don't have any authority here, I can only offer advice how I would approach this.

What do I mean by "this"? I think the code should transition to using signed integers pervasively, everywhere by default, except the places where it needs unsigned integers (such as bit fiddling and modulo-2 arithmetic, a lot of crypto code falls into this category).

I think getting there has to be incremental, and during the transition there will probably be a net increase of signed/unsigned casts. But in the end the whole codebase could be built with `-Wsign-conversion -Werror`, which would be a great place to be in, and with fewer casts.

@practicalswift I believe you've found the issues you've addressed here with `-fsanitize=unsigned`, right? I think that's a good tool to find the most egregious misuses of unsigned integers, they should probably be the first ones to go. However, perhaps addressing all of these individual spots in a single commit isn't the right slicing or increment towards the goal of reducing misuses of unsigned integers. Perhaps instead, pick one, and fix that one properly, including all the ripple-effects it may have.

for example, take [this](https://github.com/bitcoin/bitcoin/pull/11535/files#diff-ca74c4b28865382863b8fe7633a85cd6R219) part of the patch.

Maybe it would make more sense to explore what would happen if `GetTxSize()` would return an `int64_t` (which seems like a cleaner solution than casting its result like this).

**Comment by practicalswift** [comment]
@arvidn "This" was referring to this PR - more specifically getting rid of unintended unsigned wraparounds in the code base.

@arvidn @MarcoFalke Are there any parts of this PR that you can ACK or NACK? I can limit this PR if required to get some progress :-)

File: src/validation.cpp [hunk 13]
```diff
@@ -4707,7 +4707,8 @@ bool LoadMempool()
         }
         uint64_t num;
         file >> num;
-        while (num--) {
+        while (num) {
+            --num;
```
Comment (ken2812221) [inline src/validation.cpp]: How do this avoid overflow?

Comment (sipa) [reply src/validation.cpp]: It won't decrement when `num` is 0. The old code would stop the loop, but still decrement.
